### ISSUE TITLE
pytest failure: shutil.py:606: PermissionError, bump.py:126: SystemExit

### ISSUE THREAD

**Issue by gabe-camp** [Woile/commitizen#67]
Env:
- windows 10
- python38-32/python.exe
- pycharm 2018.3.1 (community)
- git bash terminal

Repro Steps:
1. $ pip install poetry
2. $ poetry install
3. $ poetry run pytest

Issue:

ERROR at teardown of test_bump_command : 
..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:606: PermissionError
..\..\commitizen\commands\bump.py:126: SystemExit


pytest output:

=========================================================================================================================== test session starts ===========================================================================================================================
platform win32 -- Python 3.8.0, pytest-5.2.4, py-1.8.0, pluggy-0.13.0
rootdir: C:\Users\campanellag\workspace\commitizen
plugins: cov-2.8.1, mock-1.11.2
collected 114 items                                                                                                                                                                                                                                                        

tests\test_bump_command.py FE                                                                                                                                                                                                                                        [  0%]
tests\test_bump_create_commit_message.py ...                                                                                                                                                                                                                         [  3%]
tests\test_bump_create_tag.py ......                                                                                                                                                                                                                                 [  8%]
tests\test_bump_find_increment.py .......                                                                                                                                                                                                                            [ 14%]
tests\test_bump_find_version.py ..........................................                                                                                                                                                                                           [ 51%]
tests\test_bump_update_version_in_files.py ..                                                                                                                                                                                                                        [ 53%]
tests\test_cli.py ...                                                                                                                                                                                                                                                [ 56%]
tests\test_commands.py ..........                                                                                                                                                                                                                                    [ 64%]
tests\test_conf.py ................                                                                                                                                                                                                                                  [ 78%]
tests\test_cz_base.py ......                                                                                                                                                                                                                                         [ 84%]
tests\test_cz_conventional_commits.py ...........                                                                                                                                                                                                                    [ 93%]
tests\test_cz_jira.py .....                                                                                                                                                                                                                                          [ 98%]
tests\test_factory.py ..                                                                                                                                                                                                                                             [100%]

================================================================================================================================= ERRORS ==================================================================================================================================
_________________________________________________________________________________________________________________ ERROR at teardown of test_bump_command __________________________________________________________________________________________________________________

    @pytest.fixture
    def create_project():
        current_directory = os.getcwd()
        tmp_proj_path = "tests/tmp-proj"
        full_tmp_path = os.path.join(current_directory, tmp_proj_path)
        if not os.path.exists(full_tmp_path):
            os.makedirs(full_tmp_path)
        os.chdir(full_tmp_path)
        yield
        os.chdir(current_directory)
>       shutil.rmtree(full_tmp_path)

tests\test_bump_command.py:23:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:730: in rmtree
    return _rmtree_unsafe(path, onerror)
..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:603: in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:603: in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:603: in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:608: in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path = 'C:\\Users\\campanellag\\workspace\\commitizen\\tests/tmp-proj\\.git\\objects\\08', onerror = <function rmtree.<locals>.onerror at 0x046C3658>

    def _rmtree_unsafe(path, onerror):
        try:
            with os.scandir(path) as scandir_it:
                entries = list(scandir_it)
        except OSError:
            onerror(os.scandir, path, sys.exc_info())
            entries = []
        for entry in entries:
            fullname = entry.path
            if _rmtree_isdir(entry):
                try:
                    if entry.is_symlink():
                        # This can only happen if someone replaces
                        # a directory with a symlink after the call to
                        # os.scandir or entry.is_dir above.
                        raise OSError("Cannot call rmtree on a symbolic link")
                except OSError:
                    onerror(os.path.islink, fullname, sys.exc_info())
                    continue
                _rmtree_unsafe(fullname, onerror)
            else:
                try:
>                   os.unlink(fullname)
E                   PermissionError: [WinError 5] Access is denied: 'C:\\Users\\campanellag\\workspace\\commitizen\\tests/tmp-proj\\.git\\objects\\08\\a40f553fb70fbd1729b27c4ef716ccbc1bbdaa'

..\..\AppData\Local\Programs\Python\Python38-32\lib\shutil.py:606: PermissionError
-------------------------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------------------------
bump: version 0.1.0 → 0.2.0
tag to create: 0.2.0
increment detected: MINOR
-------------------------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------------------------
warning: CRLF will be replaced by LF in pyproject.toml.
The file will have its original line endings in your working directory.

================================================================================================================================ FAILURES =================================================================================================================================
____________________________________________________________________________________________________________________________ test_bump_command ____________________________________________________________________________________________________________________________

mocker = <pytest_mock.plugin.MockFixture object at 0x046C1C10>, create_project = None

    def test_bump_command(mocker, create_project):
        with open("./pyproject.toml", "w") as f:
            f.write("[tool.commitizen]\n" 'version="0.1.0"')
    
        cmd.run("git init")
    
        # MINOR
        create_file_and_commit("feat: new file")
    
        testargs = ["cz", "bump", "--yes"]
        mocker.patch.object(sys, "argv", testargs)
>       cli.main()

..\test_bump_command.py:45:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\commitizen\cli.py:148: in main
    args.func(conf, vars(args))()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <commitizen.commands.bump.Bump object at 0x046D4A48>

    def __call__(self):
        """Steps executed to bump."""
        try:
            current_version_instance: Version = Version(self.parameters["version"])
        except TypeError:
            out.error("[NO_VERSION_SPECIFIED]")
            out.error("Check if current version is specified in config file, like:")
            out.error("version = 0.4.3")
            raise SystemExit(NO_VERSION_SPECIFIED)
    
        # Initialize values from sources (conf)
        current_version: str = self.config["version"]
        tag_format: str = self.parameters["tag_format"]
        bump_commit_message: str = self.parameters["bump_message"]
        current_tag_version: str = bump.create_tag(
            current_version, tag_format=tag_format
        )
        files: list = self.parameters["files"]
        dry_run: bool = self.parameters["dry_run"]
    
        is_yes: bool = self.arguments["yes"]
        prerelease: str = self.arguments["prerelease"]
        increment: Optional[str] = self.arguments["increment"]
    
        is_initial = self.is_initial_tag(current_tag_version, is_yes)
        commits = git.get_commits(current_tag_version, from_beginning=is_initial)
    
        # No commits, there is no need to create an empty tag.
        # Unless we previously had a prerelease.
        if not commits and not current_version_instance.is_prerelease:
            out.error("[NO_COMMITS_FOUND]")
            out.error("No new commits found.")
            raise SystemExit(NO_COMMITS_FOUND)
    
        if increment is None:
            bump_pattern = self.cz.bump_pattern
            bump_map = self.cz.bump_map
            if not bump_map or not bump_pattern:
                out.error(f"'{self.config['name']}' rule does not support bump")
                raise SystemExit(NO_PATTERN_MAP)
            increment = bump.find_increment(
                commits, regex=bump_pattern, increments_map=bump_map
            )
    
        # Increment is removed when current and next version
        # are expected to be prereleases.
        if prerelease and current_version_instance.is_prerelease:
            increment = None
    
        new_version = bump.generate_version(
            current_version, increment, prerelease=prerelease
        )
        new_tag_version = bump.create_tag(new_version, tag_format=tag_format)
        message = bump.create_commit_message(
            current_version, new_version, bump_commit_message
        )
    
        # Report found information
        out.write(message)
        out.write(f"tag to create: {new_tag_version}")
        out.write(f"increment detected: {increment}")
    
        # Do not perform operations over files or git.
        if dry_run:
            raise SystemExit()
    
        config.set_key("version", new_version.public)
        bump.update_version_in_files(current_version, new_version.public, files)
        c = git.commit(message, args="-a")
        if c.err:
            out.error(c.err)
>           raise SystemExit(COMMIT_FAILED)
E           SystemExit: 8

..\..\commitizen\commands\bump.py:126: SystemExit
-------------------------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------------------------
bump: version 0.1.0 → 0.2.0
tag to create: 0.2.0
increment detected: MINOR
-------------------------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------------------------
warning: CRLF will be replaced by LF in pyproject.toml.
The file will have its original line endings in your working directory.

============================================================================================================================ warnings summary =============================================================================================================================
tests/test_cli.py::test_version
  C:\Users\campanellag\workspace\commitizen\commitizen\cli.py:133: UserWarning: 'cz --version' will be deprecated in next major version. Please use 'cz version' command from your scripts
    warnings.warn(

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=========================================================================================================== 1 failed, 113 passed, 1 warnings, 1 error in 1.84s ============================================================================================================

**Comment by Woile** [comment]
Hello, unfortunately I don't have a windows machine. I think the error can be approached with help from this SO answers: [answer 1](https://stackoverflow.com/questions/2656322/shutil-rmtree-fails-on-windows-with-access-is-denied) or [answer 2](https://stackoverflow.com/questions/1213706/what-user-do-python-scripts-run-as-in-windows). 

The second issue is not clear enoguh as of why it's failing :/ could you idenify what is the output of `out.error(c.err)` before raising `raise SystemExit(COMMIT_FAILED)`? maybe that helps

**Comment by gabe-camp** [comment]
Added the handler from [answer 2](https://stackoverflow.com/questions/1213706/what-user-do-python-scripts-run-as-in-windows) to test_bump_command.py.  Issue resolved.

The second issue was caused by a local git config issue, as in this [SO](https://stackoverflow.com/questions/1967370/git-replacing-lf-with-crlf).  Said update to my local config resolved issue.  As part of this I prettied up the error report in bump.py for c.err.
